client/src/main/java/io/pravega/client/stream/notifications/notifier/ScaleEventNotifier.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.stream.notifications.notifier;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import io.pravega.client.state.StateSynchronizer;
+import io.pravega.client.stream.impl.ReaderGroupState;
+import io.pravega.client.stream.notifications.Listener;
+import io.pravega.client.stream.notifications.NotificationSystem;
+import io.pravega.client.stream.notifications.events.ScaleEvent;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class ScaleEventNotifier extends AbstractEventNotifier<ScaleEvent> {
+
+    private static final int UPDATE_INTERVAL_SECONDS = Integer.parseInt(
+            System.getProperty("pravega.client.scaleEvent.poll.interval.seconds", String.valueOf(120)));
+    private final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier;
+    private final AtomicBoolean pollingStarted = new AtomicBoolean();
+    @GuardedBy("$lock")
+    private ScheduledFuture<?> future;
+    @GuardedBy("$lock")
+    private StateSynchronizer<ReaderGroupState> synchronizer;
+
+    public ScaleEventNotifier(final NotificationSystem notifySystem,
+                              final Supplier<StateSynchronizer<ReaderGroupState>> synchronizerSupplier,
+                              final ScheduledExecutorService executor) {
+        super(notifySystem, executor);
+        this.synchronizerSupplier = synchronizerSupplier;
+    }
+
+    @Override
+    @Synchronized
+    public void registerListener(final Listener<ScaleEvent> listener) {
+        notifySystem.addListeners(getType(), listener, this.executor);
+        //periodically fetch the scale
+        if (!pollingStarted.getAndSet(true)) { //schedule the  only once
+            synchronizer = synchronizerSupplier.get();
+            future = executor.scheduleAtFixedRate(this::checkAndTriggerScaleNotification, 0,
```

Comment:
This is not exactly what we are after. This logic is checking that the number of segments and readers remains the same, but what we need is an indicator that a stream scaling event happened. When a reader hits the end of a segment (because it is sealed) and gets successors, that's when we should trigger a notification.

It makes sense to report back the current number of readers in the group in the case the application is getting this event in a single place and wants to know how many readers are actually running.

Reply:
On a second look, this is actually polling the state of the synchronizer, which is updated when a reader hits the end of a segment. Given that the synchronizer doesn't have a notification mechanism, we need to poll, so it does sound fine.